### PR TITLE
ROB : fix infinite loop due to Invalid object

### LINKED ISSUE #1329: Infinite loop while reading metadata
When I try to read the metadata of [Effective Java 3rd Edition by Joshua Bloch.pdf](https://github.com/bookyue/my_books/blob/53e6378b740807dc7d18f871acbf80ee72962d31/programming_language/java/Effective%20Java%203rd%20Edition%20by%20Joshua%20Bloch.pdf) it takes extremely long. It might even be an infinite loop.

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Linux-5.4.0-125-generic-x86_64-with-debian-bullseye-sid

$ python -c "import PyPDF2;print(PyPDF2.__version__)"
2.10.5
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from PyPDF2 import PdfReader
import time

t0 = time.time()
reader = PdfReader("Effective Java 3rd Edition by Joshua Bloch.pdf")
t1 = time.time()
print(f"Creating PDF Reader: {t1-t0:.2f} seconds")

t0 = time.time()
metadata = reader.metadata
t1 = time.time()
print(f"Reading metadata: {t1-t0:.2f} seconds")
```

The PDF: [Effective Java 3rd Edition by Joshua Bloch.pdf](https://github.com/bookyue/my_books/blob/53e6378b740807dc7d18f871acbf80ee72962d31/programming_language/java/Effective%20Java%203rd%20Edition%20by%20Joshua%20Bloch.pdf)

### PR DESCRIPTION
**Pull request by pubpub-zz**
fixes #1329
*prevent loop within dictionnaries where objects not respecting standard *fix cmap warnings due to "numbered" characters ( #2d instead of -) *apply unnumbering to nameobject
*add _get_indirect_object for debug/dev purpose
*add some missing seeks (no issue reported yet)

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@MartinThoma 
ready for review

File: PyPDF2/generic/_data_structures.py [hunk 1]
```diff
@@ -812,10 +820,9 @@ def read_object(
 ) -> Union[PdfObject, int, str, ContentStream]:
     tok = stream.read(1)
     stream.seek(-1, 1)  # reset to start
-    idx = ObjectPrefix.find(tok)
-    if idx == 0:
+    if tok == b"/":
         return NameObject.read_from_stream(stream, pdf)
-    elif idx == 1:
+    elif tok == b"<":
```
Comment (MartinThoma) [inline PyPDF2/generic/_data_structures.py:825]: Nice that you got rid of ObjectPrefix / idx. It's way easier to read now :+1:

**Comment by MartinThoma** [comment]
I'll start my big test suite + release today :-) Thank you for taking care of the issue :pray:

**Comment by MartinThoma** [comment]
I think I should create a security advisory for this

**Pull request merged**